### PR TITLE
Add FreeBSD Travis

### PR DESCRIPTION
**Pull request by nurupo**
Just experimenting running FreeBSD in qemu and sshing into it.

This PR is not finished and is not ready for review yet.

---
This change is [Reviewable](https://reviewable.io/reviews/toktok/c-toxcore/505)

### CODE REVIEW

**Comment by nurupo** [comment]
I think it's ready to review now.

Just a few notes:

- The FreeBSD build currently fails because of the network headers issue which https://github.com/TokTok/c-toxcore/pull/473 fixes. After the fixes are merged, toxcore build successfully, but tests are failing. Do we want to put the FreeBSD build into Travis's Allowed Failures build group?

- Currently, on every build job run, we try to update the FreeBSD system and its packages to the latest version and then store the updated VM image in the Travis cache. The run time of the build job can be reduced by about 6 minutes if we don't update the system at all, apart from the initial setup.

**Comment by iphydf** [comment]
Don't put it in the allowed failures group. Instead, put `­­|| true` after the tests on freebsd. That way, we know it builds, even though the tests fail.

**Comment by Zer0-One** [comment]
@nurupo if you can find a way to create an updated image offline, and just fetch/run it, you can avoid having to do updates when travis runs.

**Comment by nurupo** [comment]
@iphydf This should do it?

Also rebased on master.

**Comment by GrayHatter** [comment]
rebuild when the toxcore version changes?

**Comment by nurupo** [comment]
Rebuild what? Not sure what you are asking about.

**Comment by GrayHatter** [comment]
> The run time of the build job can be reduced by about 6 minutes if we don't update the system at all,

**Comment by nurupo** [comment]
Sounds like a good idea, given we keep releasing often, such that the image won't get very out of date.

**Comment by iphydf** [comment]
Please enable the checkbox "Allow edits from maintainers." on the bottom right.

**Comment by nurupo** [comment]
How about updating the FreeBSD system/packages every time there is a change in git tags? It's somewhat similar to updating it every time toxcore version changes.

**Comment by nurupo** [comment]
@iphydf I will do the checkbox thing after I'm done with making the VM image update on git tag change.

**Comment by nurupo** [comment]
@iphydf clang's mutex analyzer in FreeBSD's build is complaining about mutexes.

Also, I'm not sure why hstox build keeps failing. It builds just fine in master, and I have rebased this PR on master. Did I accidentally break it in this PR?

**Comment by iphydf** [comment]
@nurupo hstox should work after #449 is merged.

**Comment by nurupo** [comment]
Restarted the FreeBSD build, as Travis got a bit messed up and showed log for one of the previous builds instead. Now you can see that it fails to build due to some static analysis on mutexes issue.

**Comment by nurupo** [comment]
It started failing on Travis out of a blue, due to no output in 10 minutes, which wasn't happening before, and it's such a pain to fix it, have been trying for a couple days now :\ Thinking that it's due to the pager that doesn't terminate, so I have put

>'echo "export PAGER=cat"             >> ~/.profile'

but apparently default FreeBSD shell doesn't read `.profile`? Now I have prefixed the problematic command with `env PAGER=cat` directly in-place, will see if that helps...

**Comment by nurupo** [comment]
Ok, the FreeBSD script seems to work correctly, the build now fails due to the compiler erroring, so at least some progress. Next up is to fix the compilation error and hope there is no more of those.

**Comment by nurupo** [comment]
Squashed commits, rebased on master, should be ready to review @iphydf.

**Comment by nurupo** [comment]
Toxcore tests fails horribly though, so I wouldn't expect toxcore to work on FreeBSD (how did someone manage to make ports of toxcore on FreeBSD?), so the next step would be for someone to fix toxcore so that the tests pass. Travis is set to succeed on failed FreeBSD tests not to disrupt other PRs.

**Comment by nurupo** [comment]
I could give a little refactor to the bash script, I sometimes specify flags after arguments, which is not the correct order, but then I'm not sure which commit I would have to squash that change into, and it would probably conflict with other commits I have. I could make it as a separate commit, I guess.

**Comment by thierry-FreeBSD** [comment]
@nurupo : the FreeBSD port is lagging to release 0.1.8, 0.1.9 is not portable enough and would require too much patches.

**Comment by robinlinden** [comment]
:lgtm_strong:

---

Reviewed 2 of 5 files at r1, 1 of 1 files at r4, 3 of 3 files at r5, 4 of 4 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/505#-:-KqnVQDshjsvc9VhY2sP:b-jbsf8z)*

**Comment by nurupo** [comment]
@thierry-FreeBSD do you patch 0.1.7 in order to make it work or does it work out of the box? Would be interesting to see the patches if you do patch it.

**Comment by thierry-FreeBSD** [comment]
@nurupo : no patches ATM. When I commit patches to the ports tree, they are viewable from
https://svnweb.freebsd.org/ports/head/net-im/tox/files/

**Comment by nurupo** [comment]
@iphydf any day now

**Comment by endoffile78** [comment]
:lgtm_strong:

---

Reviewed 2 of 5 files at r1, 3 of 3 files at r5, 3 of 4 files at r6, 2 of 2 files at r7.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/505#-:-Krs2NilWSpzQvDElNHj:b-jbsf8z)*

File: other/travis/freebsd-install [hunk 1]
```diff
@@ -0,0 +1,263 @@
+#!/bin/sh
+
+# Travis doesn't provide FreeBSD machines, so we just take a Linux one and run
+# FreeBSD in qemu virtual machine. qemu is being ran in curses mode inside a
+# screen session, because screen allows to easily send input and read output.
+# The input is sent using `screen -S session-name -X stuff ...` and the output
+# is read from the screen's log file. Note that for some reason you can't send
+# long input lines on Travis (it works just fine when I do it on my machine...),
+# but that limitation is not an issue, as we don't really need to send long
+# lines of input anyway. Also, note that sice we run qemu in curses mode, the
```
Comment (iphydf) [inline other/travis/freebsd-install]: s/sice/since

File: other/travis/freebsd-install [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+# output contains control characters intended for a terminal emulator telling
+# how to position and color the text, so it might be a little tricky to read it
+# sometimes. The only time when this script has to send input to and read the
+# output from the screen session is during the initial setup when we setup the
+# network, install and configure the ssh server, and update the system. After
+# this initial setup, ssh is used to communicate with the FreeBSD running in the
+# VM, which is a lot friendlier way of communication. Please note that Travis
+# doesn't seem to allow KVM passthreough, so qemu has to emulate all the
```
Comment (iphydf) [inline other/travis/freebsd-install]: passthrough

File: other/travis/freebsd-install [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+# hardware, which makes it quite slow compared to the host machine. We cache
+# the qemu image since it takes a long time to run the initial system and
+# pacakge updates, and we do incremental system and package updates on every
```
Comment (iphydf) [inline other/travis/freebsd-install]: package

File: other/travis/freebsd-install [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+  execute_shell_and_wait 'echo "PasswordAuthentication yes" >> /etc/ssh/sshd_config'
+  execute_shell_and_wait 'echo "PermitEmptyPasswords yes"   >> /etc/ssh/sshd_config'
+  execute_shell_and_wait 'echo "sshd_enable=YES"            >> /etc/rc.conf'
+  send_keys 'sh /etc/rc.d/netif restart && sh /etc/rc.d/sshd start && passwd
+'
+
+  # Wait for the password prompt
+  wait_for "Changing local password for root"
+
+  # Reset password to empty for the paswordless ssh to work
```
Comment (iphydf) [inline other/travis/freebsd-install]: passwordless

File: other/travis/freebsd-install [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+# Blocks until a specific text appears on VM's screen
+wait_for()
+{
+  while ! grep -q "$1" screenlog.0
+  do
+    sleep 1
+  done
+}
+
+# Starts VM and waits until its fully running (until a login prompt is shown)
```
Comment (iphydf) [inline other/travis/freebsd-install]: it's

File: other/travis/freebsd-install [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+start_vm()
+{
+  rm -f screenlog.0
+
+  # Start emulator. 2000mb RAM should be enough, right? The build machine has over 7gb.
```
Comment (iphydf) [inline other/travis/freebsd-install]: This is great, and it's very good to know how much RAM we're using during the build. If this increases beyond the 2G in the VM, we may be doing something dodgy.

File: other/travis/freebsd-install [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+
+  # Select the 1st option
+  send_keys '
+'
+
+  # Wait for the system to boot and present the login prompt
+  wait_for "FreeBSD/amd64 ("
+}
+
+# Shutdowns VM and waits until its process finishes
```
Comment (iphydf) [inline other/travis/freebsd-install]: "Shuts down" or "Shuts VM down"

File: other/travis/freebsd-install [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+
+  unxz $IMAGE_NAME.xz
+
+  # With this we don't have to guess how long a command will run for and sleeping
+  # for that amount of time, risking either under sleeping or over sleeping, instead
+  # we will sleep exactly until the command is finished by printing out a unique
+  # string after the command is executed and then checking if it was printed.
+  execute_shell_and_wait()
+  {
+    # $RANDOM is a bash built-in, so we try to avoid name collosion here by using ugly RANDOM_STR name
```
Comment (iphydf) [inline other/travis/freebsd-install]: collision

**Comment by nurupo** [comment]
Fixed typos.

**Comment by robinlinden** [comment]
Reviewed 2 of 2 files at r7, 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/505)*

**Pull request merged**